### ISSUE TITLE
PermissionError trying to make first deployment

### ISSUE THREAD

**Issue by rgov** [Miserlou/Zappa#1295]
## Context
I'm trying to do my first Zappa deployment and hit an exception:

    Calling deploy for stage dev..
    Creating project-dev-ZappaLambdaExecutionRole IAM Role..
    Creating zappa-permissions policy on project-dev-ZappaLambdaExecutionRole IAM Role.
    Oh no! An error occurred! :(
    
    ==============
    
    Traceback (most recent call last):
      File "/.../.direnv/python-3.6.3/lib/python3.6/site-packages/zappa/cli.py", line 2610, in handle
        sys.exit(cli.handle())
      File "/.../.direnv/python-3.6.3/lib/python3.6/site-packages/zappa/cli.py", line 505, in handle
        self.dispatch_command(self.command, stage)
      File "/.../.direnv/python-3.6.3/lib/python3.6/site-packages/zappa/cli.py", line 539, in dispatch_command
        self.deploy(self.vargs['zip'])
      File "/.../.direnv/python-3.6.3/lib/python3.6/site-packages/zappa/cli.py", line 709, in deploy
        self.create_package()
      File "/.../.direnv/python-3.6.3/lib/python3.6/site-packages/zappa/cli.py", line 2171, in create_package
        disable_progress=self.disable_progress
      File "/.../.direnv/python-3.6.3/lib/python3.6/site-packages/zappa/core.py", line 504, in create_lambda_zip
        copytree(cwd, temp_project_path, symlinks=False, ignore=shutil.ignore_patterns(*excludes))
      File "/.../.direnv/python-3.6.3/lib/python3.6/site-packages/zappa/utilities.py", line 54, in copytree
        copytree(s, d, symlinks, ignore)
      File "/.../.direnv/python-3.6.3/lib/python3.6/site-packages/zappa/utilities.py", line 54, in copytree
        copytree(s, d, symlinks, ignore)
      File "/.../.direnv/python-3.6.3/lib/python3.6/site-packages/zappa/utilities.py", line 54, in copytree
        copytree(s, d, symlinks, ignore)
      File "/.../.direnv/python-3.6.3/lib/python3.6/site-packages/zappa/utilities.py", line 32, in copytree
        shutil.copystat(src, dst)
      File "/.../.direnv/python-3.6.3/lib/python3.6/shutil.py", line 218, in copystat
        lookup("chflags")(dst, st.st_flags, follow_symlinks=follow)
    PermissionError: [Errno 1] Operation not permitted: '/var/folders/dy/0lgxvt4n0nb4c6h0sgq66bl80000gn/T/1513418744/.venv/include/python2.7' 

## Your Environment
* Python environment managed by `direnv`
* macOS 10.13.2
* Zappa version used: 0.45.1
* Operating System and Python version: 3.6.3
* The output of `pip freeze`:

```
argcomplete==1.9.2
base58==0.2.4
boto3==1.5.1
botocore==1.8.15
certifi==2017.11.5
cfn-flip==1.0.0
chardet==3.0.4
click==6.7
defusedxml==0.5.0
dj-database-url==0.4.2
Django==2.0
docutils==0.14
durationpy==0.5
future==0.16.0
gunicorn==19.7.1
hjson==3.0.1
idna==2.6
jmespath==0.9.3
kappa==0.6.0
lambda-packages==0.19.0
oauthlib==2.0.6
placebo==0.8.1
psycopg2==2.7.3.1
PyJWT==1.5.3
python-dateutil==2.6.1
python-slugify==1.2.4
python3-openid==3.1.0
pytz==2017.3
PyYAML==3.12
requests==2.18.4
requests-oauthlib==0.8.0
s3transfer==0.1.12
six==1.11.0
social-auth-app-django==2.0.0
social-auth-core==1.5.0
toml==0.9.3
tqdm==4.19.1
troposphere==2.1.2
Unidecode==0.4.21
urllib3==1.22
Werkzeug==0.12
whitenoise==3.3.1
wsgi-request-logger==0.4.6
zappa==0.45.1
```

* Your `zappa_settings.py`: 

```
{
    "dev": {
        "aws_region": null,
        "django_settings": "mysite.settings",
        "profile_name": "default",
        "project_name": "project",
        "runtime": "python3.6",
        "s3_bucket": "xxxxx"
    }
}
```

**Comment by rgov** [comment]
Strange that it's doing something with this path, which (a) isn't in my `.direnv` directory and (b) references Python 2.7, which is not the version I'm using.

`/var/folders/dy/0lgxvt4n0nb4c6h0sgq66bl80000gn/T/1513418744/.venv/include/python2.7`

**Comment by rgov** [comment]
```
temp_project_path = os.path.join(tempfile.gettempdir(), str(int(time.time())))
```

This isn't a good way to generate a temporary directory name. It should use `tempfile.mkdtemp`.

**Comment by rgov** [comment]
Here are the parameters to `copyfile` that are failing:

```
copytree()
  src =  /.../.venv/include/python2.7
  dst =  /var/folders/dy/0lgxvt4n0nb4c6h0sgq66bl80000gn/T/tmp107yyxhz/.venv/include/python2.7
  symlinks =  False
  ignore =  <function ignore_patterns.<locals>._ignore_patterns at 0x108c91048>
```

In this case, `/.../.venv/include/python2.7` is a symlink to `/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7`. When `shutil.copy2` tries to copy the ownership data from the system directory to the new copy, it cannot, because the current user does not have permission to transfer files to the root user.

**Comment by rgov** [comment]
Well, that theory must be close, but not quite:

> The file contents, owner, and group are unaffected

Ok, turning to `sudo dtruss -f sudo -u $USER python /tmp/script.py` on my reduced test case:

```
40797/0x221588:  chflags("/tmp/dst/classobject.h\0", 0x80020, 0x0)		 = -1 Err#1
```

According to [stat.h](https://opensource.apple.com/source/xnu/xnu-4570.1.46/bsd/sys/stat.h.auto.html) `0x80020 = SF_RESTRICTED | UF_COMPRESSED`.

`SF_RESTRICTED` can only be set by the superuser and is related to System Integrity Protection. We don't want to copy this flag.

I have a patch which is not very graceful but addresses the issue.

**Comment by rgov** [comment]
https://bugs.python.org/issue32347

**Comment by gdudek** [comment]
I observed this bug in issue #1098 and suggested an ugly short-term work-around that people can use until this fix is accepted.  [https://github.com/Miserlou/Zappa/issues/1098#issuecomment-330257054]

**Comment by scoates** [comment]
Looks like this was fixed in #1297. Please reopen if I'm incorrect.
